lib/gc/Dialect/Linalgx/LinalgxOps.cpp

```diff
@@ -82,6 +82,385 @@ void SigmoidOp::getEffects(
+                     .getValue());
+  maps.back() = simplifyAffineMap(
+      maps.back().replaceDimsAndSymbols({}, symbolBindings, 6, 0));
+  cached = Builder(context).getAffineMapArrayAttr(maps);
+  getOperation()->setAttr(memoizeAttr, cached);
+  return cached;
+}
+
+void Mmt2DVnniOp::regionBuilder(ImplicitLocOpBuilder &b, Block &block,
+                                ArrayRef<NamedAttribute> attrs) {
+  assert(3 > 0 && block.getNumArguments() == 3 &&
+         "Mmt2DVnniOp regionBuilder expects 3 (>=0) args");
+  RegionBuilderHelper helper(b, block);
+  SmallVector<Value> yields;
+
+  Value value1 =
+      helper.buildTypeFn(TypeFn::cast_signed, block.getArgument(2).getType(),
+                         block.getArgument(0));
+  Value value2 =
+      helper.buildTypeFn(TypeFn::cast_signed, block.getArgument(2).getType(),
+                         block.getArgument(1));
+  Value value3 = helper.buildBinaryFn(BinaryFn::mul, value1, value2);
+  Value value4 =
+      helper.buildBinaryFn(BinaryFn::add, block.getArgument(2), value3);
+  yields.push_back(value4);
+  helper.yieldOutputs(yields);
+}
+
+ParseResult Mmt2DVnniOp::parse(OpAsmParser &parser, OperationState &result) {
+  return ::parseNamedStructuredOp(parser, result,
+                                  Mmt2DVnniOp::getNumRegionArgs(),
+                                  Mmt2DVnniOp::getRegionBuilder());
+}
+
+void Mmt2DVnniOp::print(OpAsmPrinter &p) {
+  ::printNamedStructuredOp(p, getOperation(), getInputs(), getOutputs());
+}
+
+LogicalResult Mmt2DVnniOp::fold(FoldAdaptor, SmallVectorImpl<OpFoldResult> &) {
+  return memref::foldMemRefCast(*this);
+}
+
+void Mmt2DVnniOp::getEffects(
+    SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+        &effects) {
+  if (hasPureTensorSemantics())
+    return;
+  getGenericEffectsImpl(effects, getOperation()->getResults(), getDpsInputs(),
+                        getDpsInits());
+}
+
+LogicalResult Mmt2DVnniOp::verify() {
+  // A[M, K]
+  // B[N0, K0, K1, N1, K2]
+  // C[M, N]
+  auto shapeA = ShapeAdaptor(getInputs()[0].getType());
+  auto shapeB = ShapeAdaptor(getInputs()[1].getType());
+  auto shapeC = ShapeAdaptor(getOutputs()[0].getType());
+  // check rank
+  auto hasRank = shapeA.hasRank() && shapeB.hasRank() && shapeC.hasRank();
+  if (!hasRank)
+    return emitOpError() << "input/output must have rank.";
+  auto checkRank = (shapeA.getRank() == 2) && (shapeB.getRank() == 5) &&
+                   (shapeC.getRank() == 2);
+  if (!checkRank)
+    return emitOpError() << "not supported input/output shape.";
+  // match M, N, K dims
+  bool matchM = shapeA.getDimSize(0) == shapeC.getDimSize(0);
+  bool matchN =
+      (shapeB.getDimSize(0) * shapeB.getDimSize(3)) == shapeC.getDimSize(1);
+  bool matchK =
+      shapeA.getDimSize(1) ==
+      (shapeB.getDimSize(1) * shapeB.getDimSize(2) * shapeB.getDimSize(4));
+  bool matchVnni = (shapeB.getDimSize(4) == 1) || (shapeB.getDimSize(4) == 2) ||
```

Comment:
Same as above, why do we consider `1` as vnni format here?

Reply:
```
input: MKmk/MK
weight: NKkn4k/NKkn2k/NKkn/KN
output: MNmn/MN
```
For NKkn format, can we treat it as NKkn1k so we can reuse this op?

Reply:
Is there a name for `NKkn` format (differentiating from mmt4d's `Nknk`)?

Reply:
> Is there a name for `NKkn` format (differentiating from mmt4d's `Nknk`)?

The name is NKkn. for op naming, I think it's mm4d_vnni(no transpose).